ls1intum/Artemis
# Enable account registration

This PR enables the registration functionality that is offered by default by the jhipster template.
By default, it is disabled. Admins can enable it in the yml file.
They can also specify a certain email regex pattern for specific emails (e.g. @tum.de).

Fixes #1987 

### Checklist

### Motivation and Context
Allow Artemis admins to activate the default Jhipster registration functionality.

### Description
additional client side improvements:
* eliminate lots of duplicated code in quiz statistics, fix a couple of issues with respect to the charts display
* adapt new jhipster code for components in the account module
* use a constant for roles
* remove sugar as it was not needed

### Missing
- [x] Improve error handling when the registration does not work, currently the error shown to the user is generic (Will be resolved in a follow up PR)
- [x] Email pattern display and validation on client side (the one that can specified by the admin in yml)
- [x] Testing all the different cases extensively 

### Steps for Testing on **Test Server** 2
1. Make sure you are not logged in
2. Click on Register and follow the steps (email -> activation -> login)
3. Check that the user is also created on Gitlab
4. Click on Passwort Reset and follow the steps
5. Check that the changed password was also updated on Gitlab after reset
6. Try out password change (in the user menu in the upper right corner)
7. Try out settings with name changes (in the user menu in the upper right corner)
8. Please also test quiz statistics and check everything is working properly (I removed lots of duplicated code)

### Test Coverage
* Server: 
    * AccountResource: 78%
    * UserService: 81%
* Client:
    * main/webapp/app/account/activate: 81 %
    * main/webapp/app/account/password: 80 %
    * main/webapp/app/account/password-reset/finish: 96 %
    * main/webapp/app/account/password-reset/init: 87 %
    * main/webapp/app/account/register: 83 %
    * main/webapp/app/account/settings: 94 %



### Screenshots

<img width="405" alt="image" src="https://user-images.githubusercontent.com/744067/95060202-3ffbf680-06fa-11eb-93a4-bafad5e8a9c3.png">

<img width="611" alt="image" src="https://user-images.githubusercontent.com/744067/95060220-45594100-06fa-11eb-968f-ee4cb1e5e3ea.png">
